nav2_map_server/src/vo_server/vector_object_shapes.cpp

```diff
@@ -0,0 +1,579 @@
+      shape_name.c_str(), uuid_str);
+  }
+
+  return true;
+}
+
+// ---------- Polygon ----------
+
+Polygon::Polygon(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  const nav2_msgs::msg::PolygonVO::SharedPtr params)
+: Shape::Shape(node)
+{
+  type_ = POLYGON;
+
+  if (!params) {
+    params_ = std::make_shared<nav2_msgs::msg::PolygonVO>();
+  } else {
+    params_ = params;
+    checkConsistency();
+  }
+  if (!polygon_) {
+    polygon_ = std::make_shared<geometry_msgs::msg::Polygon>();
+  }
+  polygon_->points = params_->points;
+
+  // If no UUID was specified, generate a new one
+  if (uuid_is_null(params_->uuid.uuid.data())) {
+    uuid_generate(params_->uuid.uuid.data());
+  }
+}
+
+bool Polygon::getROSParameters(const std::string & shape_name)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  params_->header.frame_id = getROSParameter(
+    node, shape_name + ".frame_id", "map").as_string();
+  params_->value = getROSParameter(
+    node, shape_name + ".value", nav2_util::OCC_GRID_OCCUPIED).as_int();
+
+  params_->closed = getROSParameter(
+    node, shape_name + ".closed", true).as_bool();
+
+  std::vector<double> poly_row;
+  try {
+    poly_row = getROSParameter(
+      node, shape_name + ".points", rclcpp::PARAMETER_DOUBLE_ARRAY).as_double_array();
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(
+      node->get_logger(),
+      "Error while getting polygon %s parameters: %s",
+      shape_name.c_str(), ex.what());
+    return false;
+  }
+  // Check for points format correctness
+  if (poly_row.size() <= 6 || poly_row.size() % 2 != 0) {
+    RCLCPP_ERROR(
+      node->get_logger(),
+      "Polygon %s has incorrect points description",
+      shape_name.c_str());
+    return false;
+  }
+
+  // Obtain polygon vertices
+  geometry_msgs::msg::Point32 point;
+  bool first = true;
+  for (double val : poly_row) {
+    if (first) {
+      point.x = val;
+    } else {
+      point.y = val;
+      params_->points.push_back(point);
+    }
+    first = !first;
+  }
+
+  // Filling the polygon_ with obtained points in map's frame
+  polygon_->points = params_->points;
+
+  // Getting shape UUID
+  return getShapeUUID(shape_name, params_->uuid.uuid.data());
+}
+
+void Polygon::getBoundaries(double & min_x, double & min_y, double & max_x, double & max_y)
+{
+  min_x = std::numeric_limits<double>::max();
+  min_y = std::numeric_limits<double>::max();
+  max_x = std::numeric_limits<double>::lowest();
+  max_y = std::numeric_limits<double>::lowest();
+
+  for (auto point : polygon_->points) {
+    if (point.x < min_x) {
+      min_x = point.x;
+    }
+    if (point.y < min_y) {
+      min_y = point.y;
+    }
+    if (point.x > max_x) {
+      max_x = point.x;
+    }
+    if (point.y > max_y) {
+      max_y = point.y;
+    }
+  }
+}
+
+bool Polygon::isPointInside(const double px, const double py) const
```

Comment:
This need to be considered move to common part (e.g. to `nav2_util/include/nav2_util/vector_object_utils.hpp`), as Collision Monitor has also the similar routine

Reply:
I leave that to your discretion

Reply:
Finally moved to `nav2_util` in 6d04b15af60e7089f9ad771b77c95061ee26bad7 to avoid code duplication (fortunately, it could be easily combined with existing code by using structure templating in `isPointInsidePolygon()` function)